File: Go/benchmark/datastructures/Queue.go
package datastructures

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"os"
)

type QueueNode struct {
	data int
	next *QueueNode
}

type MyQueue struct {
	frontNode *QueueNode
	rearNode  *QueueNode
}

func NewMyQueue() *MyQueue {
	return &MyQueue{frontNode: nil, rearNode: nil}
}

func (q *MyQueue) Push(value int) {
	newNode := &QueueNode{data: value, next: nil}
	if q.rearNode == nil {
		q.frontNode = newNode
		q.rearNode = newNode
		return
	}
	q.rearNode.next = newNode
	q.rearNode = newNode
}

func (q *MyQueue) Pop() {
	if q.frontNode == nil {
		return
	}
	q.frontNode = q.frontNode.next
	if q.frontNode == nil {
		q.rearNode = nil
	}
}

func (q *MyQueue) Peek() (int, error) {
	if q.frontNode == nil {
		return 0, errors.New("queue empty")
	}
	return q.frontNode.data, nil
}

func (q *MyQueue) Print() {
	fmt.Print("Queue [")
	curr := q.frontNode
	for curr != nil {
		fmt.Print(curr.data)
		if curr.next != nil {
			fmt.Print(" -> ")
		}
		curr = curr.next
	}
	fmt.Println("]")
}

// Binary Serialization
func (q *MyQueue) Serialize(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for writing: %w", err)
	}
	defer file.Close()

	count := uint64(0)
	curr := q.frontNode
	for curr != nil {
		count++
		curr = curr.next
	}

	if err := binary.Write(file, binary.LittleEndian, count); err != nil {
		return err
	}

	curr = q.frontNode
	for curr != nil {
		if err := binary.Write(file, binary.LittleEndian, int32(curr.data)); err != nil {
			return err
		}
		curr = curr.next
	}
	return nil
}

func (q *MyQueue) Deserialize(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for reading: %w", err)
	}
	defer file.Close()

	// Clear existing queue
	for q.frontNode != nil {
		q.Pop()
	}

	var count uint64
	if err := binary.Read(file, binary.LittleEndian, &count); err != nil {
		return err
	}

	for i := uint64(0); i < count; i++ {
		var value int32
		if err := binary.Read(file, binary.LittleEndian, &value); err != nil {
			return err
		}
		q.Push(int(value))
	}
	return nil
}

// JSON Serialization
type queueJSON struct {
	Data []int `json:"data"`
}

func (q *MyQueue) SerializeJSON(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for writing: %w", err)
	}
	defer file.Close()

	data := make([]int, 0)
	curr := q.frontNode
	for curr != nil {
		data = append(data, curr.data)
		curr = curr.next
	}

	queueData := queueJSON{Data: data}
	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	return encoder.Encode(queueData)
}

func (q *MyQueue) DeserializeJSON(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for reading: %w", err)
	}
	defer file.Close()

	// Clear existing queue
	for q.frontNode != nil {
		q.Pop()
	}

	var queueData queueJSON
	decoder := json.NewDecoder(file)
	if err := decoder.Decode(&queueData); err != nil {
		return err
	}

	for _, value := range queueData.Data {
		q.Push(value)
	}
	return nil
}

File: structures/Queue.cpp
#include "Queue.h"
#include <iostream>

MyQueue::MyQueue() : frontNode(nullptr), rearNode(nullptr) {}

MyQueue::~MyQueue() { while (frontNode != nullptr) { pop(); } }

void MyQueue::push(int value) {
    auto* newNode = new QueueNode(value);
    if (rearNode == nullptr) { frontNode = rearNode = newNode; return; }
    rearNode->next = newNode;
    rearNode = newNode;
}

void MyQueue::pop() {
    if (frontNode == nullptr) { return; }
    QueueNode* temp = frontNode;
    frontNode = frontNode->next;
    if (frontNode == nullptr) { rearNode = nullptr; }
    delete temp;
}

int MyQueue::peek() const {
    if (frontNode == nullptr) { throw std::runtime_error("Queue empty"); }
    return frontNode->data;
}

void MyQueue::print() const {
    std::cout << "Queue [";
    QueueNode* curr = frontNode;
    while (curr != nullptr) {
        std::cout << curr->data;
        if (curr->next != nullptr) std::cout << " -> ";
        curr = curr->next;
    }
    std::cout << "]\n";
}

void MyQueue::serialize(const std::string& filename) const {
    std::ofstream file(filename, std::ios::binary);
    if (!file) throw std::runtime_error("Cannot open file for writing");
    
    size_t count = 0;
    QueueNode* curr = frontNode;
    while (curr != nullptr) { count++; curr = curr->next; }
    
    file.write(reinterpret_cast<const char*>(&count), sizeof(count));
    curr = frontNode;
    while (curr != nullptr) {
        file.write(reinterpret_cast<const char*>(&curr->data), sizeof(int));
        curr = curr->next;
    }
    file.close();
}

void MyQueue::deserialize(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) throw std::runtime_error("Cannot open file for reading");
    
    // Clear existing queue
    while (frontNode != nullptr) { pop(); }
    
    size_t count;
    file.read(reinterpret_cast<char*>(&count), sizeof(count));
    for (size_t i = 0; i < count; i++) {
        int value;
        file.read(reinterpret_cast<char*>(&value), sizeof(int));
        push(value);
    }
    file.close();
}
